lxd/identity/cache.go

```diff
@@ -0,0 +1,151 @@
+package identity
+
+import (
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// Cache represents an thread-safe in-memory cache of the identities in the database.
+type Cache struct {
+	entries map[string]map[string]CacheEntry
+	mu      sync.RWMutex
+}
+
+// CacheEntry represents an identity. The Certificate field is pre-computed for identities with AuthenticationMethod
+// api.AuthenticationMethodTLS.
+type CacheEntry struct {
+	Identifier           string
+	AuthenticationMethod string
+	IdentityType         string
+	Projects             []string
+	Certificate          *x509.Certificate
+}
+
+// Get returns a single CacheEntry by its authentication method and identifier.
+func (c *Cache) Get(authenticationMethod string, identifier string) (*CacheEntry, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.entries == nil {
+		c.entries = make(map[string]map[string]CacheEntry)
+	}
+
+	entriesByAuthMethod, ok := c.entries[authenticationMethod]
+	if !ok {
+		return nil, api.StatusErrorf(http.StatusNotFound, "Identity %q (%s) not found", identifier, authenticationMethod)
+	}
+
+	entry, ok := entriesByAuthMethod[identifier]
+	if !ok {
+		return nil, api.StatusErrorf(http.StatusNotFound, "Identity %q (%s) not found", identifier, authenticationMethod)
+	}
+
+	return &entry, nil
+}
+
+// GetByType returns a map of identifier to CacheEntry, where all entries have the given identity type.
+func (c *Cache) GetByType(identityType string) map[string]CacheEntry {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.entries == nil {
+		c.entries = make(map[string]map[string]CacheEntry)
+	}
+
+	var authenticationMethod string
+	switch identityType {
+	case api.IdentityTypeCertificateServer, api.IdentityTypeCertificateMetrics, api.IdentityTypeCertificateClientRestricted, api.IdentityTypeCertificateClientUnrestricted:
```

Comment:
I feel like this list should be out of the cache specific function and in a more general "method to types" map so its more obvious it needs to be maintained when adding new tls certificate types.

Reply:
Or potentially we add an auth method argument to this function as well.